tool/net/redbean.c

```diff
@@ -3686,33 +3686,48 @@ static int LuaFetch(lua_State *L) {
   bool usessl;
   uint32_t ip;
   struct Url url;
-  int t, ret, sock, method;
+  int t, ret, sock, methodidx;
   char *host, *port;
   struct TlsBio *bio;
   struct Buffer inbuf;
   struct addrinfo *addr;
   struct HttpMessage msg;
   struct HttpUnchunker u;
-  const char *urlarg, *request, *body;
-  size_t urlarglen, requestlen;
-  size_t g, i, n, hdrsize, paylen;
+  const char *urlarg, *request, *body, *method;
+  char *conlenhdr = "";
+  size_t urlarglen, requestlen, paylen, bodylen;
+  size_t g, i, n, hdrsize;
   struct addrinfo hints = {.ai_family = AF_INET,
                            .ai_socktype = SOCK_STREAM,
                            .ai_protocol = IPPROTO_TCP,
                            .ai_flags = AI_NUMERICSERV};
 
   /*
-   * Get args.
+   * Get args: url [, body | {method = "PUT", body = "..."}]
    */
   urlarg = luaL_checklstring(L, 1, &urlarglen);
-  if (lua_isnoneornil(L, 2)) {
+  if (lua_istable(L, 2)) {
+    lua_settop(L, 2); // discard any extra arguments
+    lua_getfield(L, 2, "body");
+    body = luaL_optlstring(L, -1, "", &bodylen);
+    lua_getfield(L, 2, "method");
+    // use GET by default if no method is provided
+    method = strtoupper(luaL_optstring(L, -1, kHttpMethod[kHttpGet]));
+    lua_settop(L, 2); // drop all added elements to keep the stack balanced
+  } else if (lua_isnoneornil(L, 2)) {
     body = "";
-    method = kHttpGet;
+    bodylen = 0;
+    method = kHttpMethod[kHttpGet];
   } else {
-    size_t bodylen;
     body = luaL_checklstring(L, 2, &bodylen);
-    method = kHttpPost;
+    method = kHttpMethod[kHttpPost];
   }
+  // provide Content-Length header unless it's zero and not expected
+  methodidx = GetHttpMethod(method, -1);
+  if (bodylen > 0 ||
+    !(methodidx == kHttpGet || methodidx == kHttpHead ||
+      methodidx == kHttpTrace || methodidx == kHttpDelete || methodidx == kHttpConnect))
+    conlenhdr = gc(xasprintf("Content-Length: %d\r\n", bodylen));
```

Comment:
Strictly speaking you'd want `%zu` instead of `%d` but it shouldn't be a problem unless you're planning on uploading files greater than 2gb.